app/src/main/cpp/skyline/services/timesrv/time_shared_memory.cpp

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2021 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include "core.h"
+#include "time_manager_server.h"
+
+namespace skyline::service::timesrv::core {
+    struct __attribute__((packed)) TimeSharedMemoryLayout {
+        template<typename T>
+        struct ClockContextEntry {
+            u32 updateCount;
+            u32 _pad_;
+            std::array<T, 2> context;
+        };
+
+        ClockContextEntry<SteadyClockTimePoint> standardSteadyClockContextEntry;
+        ClockContextEntry<SystemClockContext> localSystemClockContextEntry;
+        ClockContextEntry<SystemClockContext> networkSystemClockContextEntry;
+        struct __attribute__((packed)) {
+            u32 updateCount;
+            std::array<u8, 2> enabled;
+        } standardUserSystemClockAutomaticCorrectionEnabledEntry;
+    };
+    static_assert(offsetof(TimeSharedMemoryLayout, localSystemClockContextEntry) == 0x38);
+    static_assert(offsetof(TimeSharedMemoryLayout, networkSystemClockContextEntry) == 0x80);
+    static_assert(offsetof(TimeSharedMemoryLayout, standardUserSystemClockAutomaticCorrectionEnabledEntry) == 0xC8);
+
+    /**
+     * @brief Time Shared Memory uses a double buffered format that alternates writes context data, this is a helper to simplify that
+     */
+    template<typename T>
+    static void UpdateTimeSharedMemoryItem(u32 &updateCount, std::array<T, 2> &item, const T &newValue) {
+        u32 newCount{updateCount + 1};
+        item[newCount & 1] = newValue;
+        asm volatile("DMB ISHST"); // 0xA
```

Comment:
It might be better to use [**std::atomic_thread_fence**](https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) with an appropriate [**std::memory_order**](https://en.cppreference.com/w/cpp/atomic/memory_order)

Reply:
None of this generates the same ASM